secretkeylabs/xverse-core
# chore: performance improvements for account derivation

# 🔘 PR Type



# 📜 Background
This PR adds to new methods to optimize the performance of restoration of account from the seed phrase. It exposes one new method `restoreWalletWithAccountsMobile` which uses `getActiveAccountsFromRootNode` and `getWalletFromRootNode`. the idea is to get the root node once and use it where is necessary since the calculation is expensive in terms of CPU. 
It also adds the method `waitFrame`, a simple wrapper of requestAnimationFrame to let the CPU breathe between expensive operations 

Issue Link: #[ENG-2449](https://linear.app/xverseapp/issue/ENG-2449/performance-improvements-for-account-derivation)


# 🔄 Changes

- [ ] Yes, Incompatible API changes
- [x] No, Adds functionality (backwards compatible)
- [ ] No, Bug fixes (backwards compatible)

Changes:
- Modify restoreWalletWithAccounts to generate the root node and uses a for loop instead of promise all to improve performance
- Create getActiveAccountsFromRootNode which receives the root node instead of generate it
- fetchActiveAccounts now uses getActiveAccountsFromRootNode to not introduce a breaking change
- Create getWalletFromRootNode which receives the root node instead of generate it
- walletFromSeedPhrase now uses getWalletFromRootNode to not introduce a breaking change
- Unify accountType definition
- None of this changes are breaking change

Impact:
- Improves performance for the mobile app when restoring accounts
- Impacts on mobile app and web extensions

# 🧪 Testing
We should test mobile and web extension in this PR and the following flows:
- Create account 
- Create wallet
- Restore wallet
- Restore accounts
- Change network on settings
